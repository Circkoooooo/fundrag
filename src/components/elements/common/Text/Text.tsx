import { ElementFunctionComponentType } from '../../type'

const Text: ElementFunctionComponentType = () => {
	return <span>文本组件</span>
}

Text.componentName = '文本组件'

export default Text
