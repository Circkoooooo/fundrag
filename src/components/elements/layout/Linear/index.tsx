import { ElementComponentType } from '../../type'

const Linear: ElementComponentType = () => {
	return <div>linear</div>
}

Linear.componentName = '线性布局'

export default Linear
