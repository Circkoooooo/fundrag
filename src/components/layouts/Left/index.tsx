import React, { useState } from 'react'
import {
	BsChevronDoubleRight as MenuCollapseRight,
	BsChevronDoubleLeft as MenuCollapseLeft,
} from 'react-icons/bs'
import { CollapseButton, HeaderMenu, LeftPanel } from './styled'

interface LeftProps {}

const Left: React.FC<LeftProps> = () => {
	const [isCollapse, setIsCollapse] = useState(true)

	return (
		<LeftPanel collapse={isCollapse}>
			<HeaderMenu>
				<CollapseButton onClick={() => setIsCollapse(!isCollapse)}>
					{isCollapse ? (
						<MenuCollapseRight
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					) : (
						<MenuCollapseLeft
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					)}
				</CollapseButton>
			</HeaderMenu>
		</LeftPanel>
	)
}

export default Left
