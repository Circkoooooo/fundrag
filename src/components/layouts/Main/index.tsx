import React from 'react'
import { ElementComponentType } from '../../elements/type'
import { MainContainer, MainScreen } from './styled'

interface MainProps {
	renderedElements?: ElementComponentType[]
}
const Main: React.FC<MainProps> = ({ renderedElements }) => {
	return (
		<MainContainer>
			<MainScreen>
				{renderedElements?.map((RenderedElement) => {
					return <RenderedElement key={crypto.randomUUID()} />
				})}
			</MainScreen>
		</MainContainer>
	)
}

Main.defaultProps = {
	renderedElements: [],
}

export default Main
