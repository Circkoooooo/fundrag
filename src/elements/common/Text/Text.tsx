import { buildBaseProps } from '../..'
import { ElementFunctionComponentType } from '../../type'

const Text: ElementFunctionComponentType = () => {
	return <span>文本组件</span>
}

Text.defaultAppendProps = buildBaseProps('inline', '文本组件')

export default Text
