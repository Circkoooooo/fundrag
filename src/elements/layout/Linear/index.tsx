import styled from 'styled-components'
import { useRef, useState } from 'react'
import { buildBaseProps } from '../..'
import { DefaultContainerProps, ElementFunctionComponentType, LayoutBaseStyleProperties } from '../../type'
import { LayoutBaseProps } from '../types'

const LinearContainer = styled.div<LayoutBaseStyleProperties>`
	${(props) => {
		return {
			width: props.width,
			height: props.height,
			backgroundColor: props.backgroundColor,
		}
	}}
`

type LinearProps = LayoutBaseProps & {}

const Linear: ElementFunctionComponentType<DefaultContainerProps, LinearProps> = ({ children, onClickEvent, defaultStyleProperties }) => {
	const styleProperties = useRef(defaultStyleProperties)

	return (
		<LinearContainer
			{...styleProperties.current}
			onClick={() => {
				onClickEvent &&
					onClickEvent({
						type: 'container',
						styleProperties: styleProperties.current,
					})
			}}
		>
			{children}
		</LinearContainer>
	)
}

Linear.defaultAppendProps = buildBaseProps('container', '线性组件')

export default Linear
