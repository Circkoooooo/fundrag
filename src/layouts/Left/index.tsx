import React, { ReactNode, useState } from 'react'
import { BsChevronDoubleRight as MenuCollapseRight, BsChevronDoubleLeft as MenuCollapseLeft } from 'react-icons/bs'
import { PanelContainer } from '../sharedStyled'
import { CollapseButton, HeaderMenu } from './styled'

interface LeftProps {
	children: ReactNode
}

const Left: React.FC<LeftProps> = ({ children }) => {
	const [isCollapse, setIsCollapse] = useState(true)

	return (
		<PanelContainer collapse={isCollapse} defaultWidth={400} minWidth={100}>
			<HeaderMenu>
				<CollapseButton onClick={() => setIsCollapse(!isCollapse)}>
					{isCollapse ? (
						<MenuCollapseRight
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					) : (
						<MenuCollapseLeft
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					)}
				</CollapseButton>
			</HeaderMenu>
			{children}
		</PanelContainer>
	)
}

export default Left
