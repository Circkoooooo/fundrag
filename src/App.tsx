import { useState } from 'react'
import styled from 'styled-components'
import RenderElement from './components/RenderElement'
import { RenderedElementsType } from './components/RenderElement/types'
import ElementSelection from './elements/ElementSelection'
import { ClickEventAttributes, ClickEventElementType } from './elements/layout/types'
import { DefaultAppendProps, DefaultContainerProps, ElementFunctionComponentType } from './elements/type'
import Left from './layouts/Left'
import Main from './layouts/Main'
import Right, { ItemAttributes, ItemUnit } from './layouts/Right'

const Background = styled.div`
	width: 100%;
	height: 100%;
	background-color: #f0f0f0;
	display: flex;
`

export type SelectElement = {
	type: keyof ClickEventElementType
	key: string
}

function App() {
	const [appendElements, setAppendElements] = useState<RenderedElementsType<DefaultAppendProps, 'inline'>[]>([])
	const [layoutElements, setLayoutElements] = useState<RenderedElementsType<DefaultContainerProps, 'container'>[]>([])

	const [selectElement, setSelectElement] = useState<SelectElement | null>(null) //当前点击选中的元素
	const [itemAttributes, setItemAttributes] = useState<ItemAttributes[]>([] as ItemAttributes[]) //选中元素所有可修改属性的数组

	const renderElement = (element: ElementFunctionComponentType<DefaultAppendProps | DefaultContainerProps>) => {
		const elementType = element.defaultAppendProps.componentType

		if (elementType === 'inline' && selectElement?.type === 'layout' && selectElement.key !== undefined) {
			const appendElement: RenderedElementsType<DefaultAppendProps, 'inline'> = {
				key: crypto.randomUUID(),
				Element: element,
				containerKey: selectElement?.key, //寻找一个容器的key
				styleProperties: element.defaultAppendProps.defaultStyleProperties,
			}
			setAppendElements([...appendElements, appendElement])
		} else if (elementType === 'container') {
			console.log(element.defaultAppendProps)
			const layoutElement: RenderedElementsType<DefaultContainerProps, 'container'> = {
				key: crypto.randomUUID(),
				Element: element,
				styleProperties: element.defaultAppendProps.defaultStyleProperties,
			}
			setLayoutElements([...layoutElements, layoutElement])
		}
	}

	// 响应每次修改属性事件
	const onEditValue = (
		elementKey: string,
		type: 'container' | 'inline',
		attrObj: {
			itemTitle: string
			itemValue: string
			itemUnit: string
		},
		event: React.FormEvent<HTMLInputElement>,
		preItemAttributes?: ItemAttributes[]
	) => {
		const { itemTitle, itemUnit } = attrObj
		if (type === 'container') {
			const newLayoutElements = layoutElements.map((element) => {
				if (element.key === elementKey) {
					;(element.styleProperties as any)[itemTitle] = event.currentTarget.value + itemUnit
				}
				return element
			})

			if (preItemAttributes) {
				setItemAttributes(preItemAttributes)
			}
			setLayoutElements(newLayoutElements)
		}
	}

	/**
	 * 响应点击渲染过的元素后的事件
	 *
	 */
	const onPickElement = (elementKey: string, type: 'container' | 'inline', clickEventsAttribute: ClickEventAttributes) => {
		setSelectElement({
			type: clickEventsAttribute.type,
			key: elementKey,
		})

		if (clickEventsAttribute.styleProperties !== undefined) {
			const items: ItemAttributes[] = Object.entries(clickEventsAttribute.styleProperties).map((item, index) => {
				const pureValue = (item[1] as string).match(/\d+/g) //不附带属性的值
				const pureUnit = (item[1] as string).match(/\D+/g) //属性的单位

				const itemTitle = item[0]
				const itemValue = pureValue === null ? ((item[1] as string) === null ? '0' : (item[1] as string)) : pureValue[0]
				const itemUnit: ItemUnit = pureUnit === null ? 'px' : pureValue === null ? '' : (pureUnit[0] as ItemUnit)

				return {
					key: index.toString(),
					itemTitle,
					itemValue,
					itemUnit,
					editValue: (event: React.FormEvent<HTMLInputElement>, preItemAttributes?: ItemAttributes[]) => {
						onEditValue(
							elementKey,
							type,
							{
								itemTitle: item[0],
								itemValue,
								itemUnit,
							},
							event,
							preItemAttributes
						)
					},
				}
			})
			setItemAttributes(items)
		}
	}

	return (
		<Background>
			<Left>
				<ElementSelection pickElement={(element) => renderElement(element)} />
			</Left>
			<Main>
				<RenderElement
					{...{ layoutElements, appendElements }}
					onPickElement={(key: string, type: 'container' | 'inline', clickEventAttributes: ClickEventAttributes) => onPickElement(key, type, clickEventAttributes)}
				/>
			</Main>
			<Right {...{ itemAttributes, selectElement }} />
		</Background>
	)
}

export default App
