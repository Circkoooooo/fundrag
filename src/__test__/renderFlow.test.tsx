import renderer, { act } from 'react-test-renderer'
import App from '../App'
import Left from '../views/layouts/Left'
import Main from '../views/layouts/Main'
import Linear from '../views/elements/layout/Linear/Linear'
import Test from '../views/elements/common/Test/Test'
import Text from '../views/elements/common/Text/Text'

describe('ensure correct process ', () => {
	const component = renderer.create(<App />)
	const instance = component.root
	const leftPanel = instance.findByType(Left)
	const mainPanel = instance.findByType(Main)
	const elementSelection = leftPanel.props.children.props

	it('load left components', () => {
		expect(elementSelection.componentPackages).toMatchSnapshot() //左侧组件的加载
	})

	it('render component in main', () => {
		const mainChildren$1 = mainPanel.props.children.props.children
		expect(mainChildren$1.length).toBe(0)
		act(() => {
			elementSelection.onItemClick(elementSelection.componentPackages[0].components[0])
		})
		const mainChildren$2 = mainPanel.props.children.props.children
		expect(mainChildren$2.length - mainChildren$1.length).toBe(1)
	})
})

describe('ensure circulate rendering', () => {
	const component = renderer.create(<App />)
	const instance = component.root
	const leftPanel = instance.findByType(Left)
	const mainPanel = instance.findByType(Main)
	let elementSelection = leftPanel.props.children.props
	let linearComponentChildren: any = null

	it('render components in main', () => {
		//render container component
		act(() => {
			elementSelection.onItemClick(Linear)
		})
		linearComponentChildren = mainPanel.props.children.props.children[0].props.children
		expect(linearComponentChildren).toBe(undefined)
		expect(mainPanel.props.children.props.children).toMatchSnapshot()

		//render non-container component to container component
		elementSelection = leftPanel.props.children.props
		act(() => {
			elementSelection.onItemClick(Test)
		})
		linearComponentChildren = mainPanel.props.children.props.children[0].props.children
		expect(linearComponentChildren.length).toBe(1)

		//rendered components tree
		expect(mainPanel.props.children.props.children).toMatchSnapshot()

		//render non-container component to container component
		act(() => {
			elementSelection.onItemClick(Text)
		})

		linearComponentChildren = mainPanel.props.children.props.children[0].props.children
		expect(linearComponentChildren.length).toBe(2)

		expect(mainPanel.props.children.props.children).toMatchSnapshot()
	})

	it('render null', () => {
		act(() => {
			elementSelection.onItemClick(null)
		})
		linearComponentChildren = mainPanel.props.children.props.children[0].props.children
		expect(linearComponentChildren.length).toBe(2)
	})
})
