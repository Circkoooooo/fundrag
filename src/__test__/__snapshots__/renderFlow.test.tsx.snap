// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensure circlate rendering render component in main 1`] = `
Array [
  <Linear />,
]
`;

exports[`ensure circlate rendering render component in main 2`] = `
Array [
  <Linear>
    <Test />
  </Linear>,
]
`;

exports[`ensure circlate rendering render component in main 3`] = `
Array [
  <Linear>
    <Test />
    <Text />
  </Linear>,
]
`;

exports[`ensure circulate rendering render components in main 1`] = `
Array [
  <Linear />,
]
`;

exports[`ensure circulate rendering render components in main 2`] = `
Array [
  <Linear>
    <Test />
  </Linear>,
]
`;

exports[`ensure circulate rendering render components in main 3`] = `
Array [
  <Linear>
    <Test />
    <Text />
  </Linear>,
]
`;

exports[`ensure correct process  load left components 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "RealNode": [Function],
        "children": Object {
          "appendChild": [Function],
          "value": Array [],
        },
        "componentRenderConfig": Object {
          "componentName": "文本组件",
        },
        "isContainer": false,
      },
      Object {
        "RealNode": [Function],
        "children": Object {
          "appendChild": [Function],
          "value": Array [],
        },
        "componentRenderConfig": Object {
          "componentName": "测试组件",
        },
        "isContainer": false,
      },
    ],
    "packageName": "常用组件",
  },
  Object {
    "components": Array [
      Object {
        "RealNode": [Function],
        "children": Object {
          "appendChild": [Function],
          "value": Array [],
        },
        "componentRenderConfig": Object {
          "componentName": "线性组件",
        },
        "isContainer": true,
      },
    ],
    "packageName": "布局组件",
  },
]
`;
